### PR TITLE
fix quantization error

### PR DESCRIPTION
**Pull request by bkiat1123**
Fixes <https://github.com/Lightning-AI/lit-parrot/issues/18>

`torch.nn.Linear` is patched with quantized linear layer in `EmptyInit` context manager, resulting in missing keys. Solved by putting the load checkpoint inside the context manager.

### CODE REVIEW

File: generate.py [hunk 1]
```diff
@@ -124,8 +124,9 @@ def main(
     t0 = time.time()
     with EmptyInitOnDevice(device=fabric.device, dtype=dtype, quantization_mode=quantize):
         model = Parrot(config)
-    with lazy_load(checkpoint_path) as checkpoint:
-        model.load_state_dict(checkpoint)
+
+        with lazy_load(checkpoint_path) as checkpoint:
```
[the remark this answers is not in the document]
Comment (carmocca) [reply generate.py:128]: The question is why the KeyError appears. For instance, Lit-LLaMA doesn't need `strict=False`: https://github.com/Lightning-AI/lit-llama/blob/main/generate.py#L134

[the remark this answers is not in the document]
Comment (carmocca) [reply generate.py:128]: > For "llm.int8", when we load checkpoint outside EmptyInit with strict=False, the model is not loaded properly. It outputs gibberish.

I tried indenting and unindenting and the output looks the same to me.

[the remark this answers is not in the document]
Comment (ArturK-85) [reply generate.py:128]: With load checkpoint inside EmptyInit with strict=False it looks work properly
Here is my output:

`Loading model 'checkpoints/stabilityai/stablelm-base-alpha-3b/lit_model.pth' with {'block_size': 4096, 'vocab_size': 50254, 'padding_multiple': 512, 'padded_vocab_size': 50688, 'n_layer': 16, 'n_head': 32, 'n_embd': 4096, 'rotary_percentage': 0.25, 'parallel_residual': True}
Time to load model: 6.52 seconds.
Global seed set to 1234
Hello, my name is Cody, and I write to you...
I need a police officer to stop shining light on troubled individuals in exchange for a lighter sentence for a crime.
I need a police officer to stop shining light on troubled individuals in exchange for a lighter sentence
Time for inference 1: 2.56 sec total, 19.54 tokens/sec
Memory used: 5.20 GB
`

[the remark this answers is not in the document]
Comment (carmocca) [reply generate.py:128]: @ArturK-85 Do you see a difference with loading inside EmptyInit vs outside? (both with `strict=False`)

Subjectively, the answer looks fine considering that these checkpoints are not very powerful.

[the remark this answers is not in the document]
Comment (ArturK-85) [reply generate.py:128]: I don't see any difference when loading inside/outside context manager both with strict=False

[the remark this answers is not in the document]
Comment (bkiat1123) [reply generate.py:128]: @carmocca 
Somehow I can't replicate difference with loading inside EmptyInit vs outside. Maybe I made some mistakes.



However, I found the root of error for `llm.int8`.

https://github.com/Lightning-AI/lit-parrot/blob/f6e17d40a4a2aa77ad4d8659411b59a0ee61c1f4/lit_parrot/quantization.py#L62-L64

When we let nn.Module load the bias. The method expects both weight and bias, but the weight has been popped off the `local_state_dict`.


Why Lit-LLaMA doesn't has this error? Because the model doesn't have bias parameters according to the `convert_hf_checkpoint` script.

[the remark this answers is not in the document]
Comment (lantiga) [reply generate.py:128]: Nice find @bkiat1123

**Comment by bkiat1123** [comment]
@carmocca @lantiga 
Summary on the issues and potential solutions.

For `llm.int8`, we get the error when we let `nn.Module` load the bias. The checkpoint loading works fine. 
We could just set `strict=False`. It has the risk of silencing any real errors. Alternatively, we could use `_load_state_dict_post_hooks` to handle the missing keys.
https://github.com/pytorch/pytorch/blob/b1bc8aecf5be5d09d76d1972a1fa1e91597bdac4/torch/nn/modules/module.py#L2032-L2040


For `gptq.int4/8`, I think the correct way is to quantize the weights, and load the quantized weights from checkpoint file?
https://github.com/Lightning-AI/lit-llama/blob/main/howto/inference.md#run-lit-llama-on-consumer-devices
"You can also use GPTQ-style int4 quantization, but this needs conversions of the weights first."

The quantized parameters have `quant_weight`, `scales`, and `zeros` attributes, instead of weights and bias. 

Porting the quantize script from lit-llama should solve the issue.

**Comment by carmocca** [comment]
Setting `strict=False` as a fix is fine with me. Thanks for your investigation @bkiat1123!

You can use this PR for it or open a separate one. But I would do int8 and gptq separately as we will also need to update the howtos and README
